use-cases/project-job-scheduling/src/main/java/org/acme/projectjobschedule/solver/ProjectJobSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,69 @@
+import org.acme.projectjobschedule.domain.JobType;
+import org.acme.projectjobschedule.domain.ResourceRequirement;
+
+public class ProjectJobSchedulingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                nonRenewableResourceCapacity(constraintFactory),
+                renewableResourceCapacity(constraintFactory),
+                totalProjectDelay(constraintFactory),
+                totalMakespan(constraintFactory)
+        };
+    }
+
+    protected Constraint nonRenewableResourceCapacity(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(ResourceRequirement.class)
+                .filter(resource -> !resource.isResourceRenewable())
+                .join(Allocation.class,
+                        Joiners.equal(ResourceRequirement::getExecutionMode, Allocation::getExecutionMode))
+                .groupBy((requirement, allocation) -> requirement.getResource(),
+                        ConstraintCollectors.sum((requirement, allocation) -> requirement.getRequirement()))
+                .filter((resource, requirements) -> requirements > resource.getCapacity())
+                .penalize(HardMediumSoftScore.ONE_HARD,
+                        (resource, requirements) -> requirements - resource.getCapacity())
+                .asConstraint("Non-renewable resource capacity");
+    }
+
+    protected Constraint renewableResourceCapacity(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(ResourceRequirement.class)
+                .filter(ResourceRequirement::isResourceRenewable)
+                .join(Allocation.class,
+                        Joiners.equal(ResourceRequirement::getExecutionMode, Allocation::getExecutionMode))
+                .flattenLast(Allocation::getBusyDates)
+                .groupBy((resourceReq, date) -> resourceReq.getResource(),
+                        (resourceReq, date) -> date,
+                        ConstraintCollectors.sum((resourceReq, date) -> resourceReq.getRequirement()))
+                .filter((resourceReq, date, totalRequirement) -> totalRequirement > resourceReq.getCapacity())
+                .penalize(HardMediumSoftScore.ONE_HARD,
+                        (resourceReq, date, totalRequirement) -> totalRequirement - resourceReq.getCapacity())
+                .asConstraint("Renewable resource capacity");
+    }
+
+    protected Constraint totalProjectDelay(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(Allocation.class)
+                .filter(allocation -> allocation.getEndDate() != null && allocation.getJobType() == JobType.SINK)
+                .impact(HardMediumSoftScore.ONE_MEDIUM, allocation -> allocation.getProjectCriticalPathEndDate() - allocation.getEndDate())
+                .asConstraint("Total project delay");
+    }
+
```

Comment:
I'm not exactly sure why we decided to make this an `impact` instead of a penalty.
Can you think why this should be a positive number?
If not, let's change the semantics and make it a penalty like anything else.

Reply:
We want to reward projects that finish before the deadline and penalize the late ones. The `impact()` function enables us to achieve that with a single constraint.

If we want only to penalize late projects, we'll need to filter them. However, it seems reasonable to reward the projects that meet the deadline.

I suggest keeping the constraint instead of penalizing late projects or creating two separate constraints. What do you think?

Reply:
Experiments have shown that changing it to a penalty constraint does not affect solution quality.